### PR TITLE
wrap end on prototype instead of instances

### PR DESCRIPTION
**Pull request by bengl**
### What does this PR do?
This avoids doing a costly Object.defineProperty on every single
instance of http.IncomingMessage passed to a server handler.

### Motivation
Object.defineProperty is a costly performance hit, so it's better to avoid doing
it where we can.

### CODE REVIEW

File: packages/dd-trace/src/plugins/util/web.js [hunk 1]
```diff
@@ -270,15 +270,26 @@ function wrapEnd (req) {
...
-    },
-    set (value) {
-      this._datadog_end = scope.bind(value, req._datadog.span)
-    }
-  })
+  res._datadog_req = req
+  res._datadog_scope = scope
+
+  if (!res._datadog_instrumented) {
+    const target = Reflect.ownKeys(res).includes('end') ? res : res.constructor.prototype
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js:277]: In which case would `res` have `end` directly if this is always expected to be the Node HTTP response object?

File: packages/dd-trace/src/plugins/util/web.js [hunk 2]
```diff
@@ -270,15 +270,26 @@ function wrapEnd (req) {
...
-    configurable: true,
-    get () {
-      return this._datadog_end
-    },
-    set (value) {
-      this._datadog_end = scope.bind(value, req._datadog.span)
-    }
-  })
+  res._datadog_req = req
+  res._datadog_scope = scope
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js:274]: Why is this needed since `scope` is basically global?

File: packages/dd-trace/src/plugins/util/web.js [hunk 3]
```diff
@@ -270,15 +270,26 @@ function wrapEnd (req) {
...
+
+  if (!res._datadog_instrumented) {
+    const target = Reflect.ownKeys(res).includes('end') ? res : res.constructor.prototype
+    Object.defineProperty(target, 'end', {
+      configurable: true,
+      get () {
+        return this._datadog_end || end
+      },
+      set (value) {
+        if (this._datadog_scope && this._datadog_req && this._datadog_req._datadog) {
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/util/web.js:284]: `this._datadog_req` is repeated a few times and is never assigned to, so it would be cleaner to move it to a const.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/plugins/util/web.js (hunk 1 above)
Comment (bengl) [reply packages/dd-trace/src/plugins/util/web.js:277]: In testing, I found that if I didn't do this, tests would fail. There are likely test cases somewhere in the suite that pass that kind of object in.

I can dig deeper to see if the test can be fixed.

File: packages/dd-trace/src/plugins/util/web.js (hunk 3 above)
Comment (bengl) [reply packages/dd-trace/src/plugins/util/web.js:284]: Sure.

File: packages/dd-trace/src/plugins/util/web.js (hunk 2 above)
Comment (bengl) [reply packages/dd-trace/src/plugins/util/web.js:274]: I was replicating the previous code which grabbed it from the local scope.

That being said, you're right. No need to stash this.

File: packages/dd-trace/src/plugins/util/web.js (hunk 1 above)
Comment (bengl) [reply packages/dd-trace/src/plugins/util/web.js]: I dug deeper. Turns out `end` was always on the instance, due to being stubbed, so this was never being set on the prototype in tests. I fixed the test object so that `end`, even when stubbed, is always in a prototype instead of an instance.

**Comment by bengl** [comment]
@rochdev Alright, all your comments have been addressed.
